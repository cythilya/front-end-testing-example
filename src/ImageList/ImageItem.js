const ImageItem = ({ title, image }) => {
  return (
    <div data-test-id="image-item">
      <img alt={title} data-test-id="image-item-src" src={image} />
      <p data-test-id="image-item-title">{title}</p>
    </div>
  );
};

export default ImageItem;
