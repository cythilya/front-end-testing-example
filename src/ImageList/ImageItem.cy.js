import ImageItem from './ImageItem';

const MOCK_IMAGE = {
    id: '1',
    title: '打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵',
    images: ['https://bit.ly/4228IT0'],
  };

describe('ImageItem', () => {
  it('should show image title correctly', () => {
    cy.mount(<ImageItem title={MOCK_IMAGE.title} image={MOCK_IMAGE.images[0]} />);

    cy.get('[data-test-id="image-item-title"]').should('have.text',
      '打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵'
    );
  });

  it('should show image url correctly', () => {
    cy.mount(<ImageItem title={MOCK_IMAGE.title} image={MOCK_IMAGE.images[0]} />);

    cy.get('[data-test-id="image-item-src"]').should('have.attr', 'src',
      'https://bit.ly/4228IT0'
    );
  });
});
