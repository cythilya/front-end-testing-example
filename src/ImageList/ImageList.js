import { useEffect, useState } from 'react';
import ImageItem from './ImageItem';

const ImageList = () => {
  const [images, setImages] = useState([]);

  useEffect(() => {
    async function fetchData() {
      const response = await fetch('https://dummyjson.com/products');

      const data = await response.json();
      setImages(data.products);
    }
    fetchData();
  }, []);

  return (
    <>
      {!images.length && (
        <div data-test-id="no-data-prompt">No data to display.</div>
      )}
      {images.map(({ id, title, images }) => (
        <ImageItem key={id} title={title} image={images[0]} />
      ))}
    </>
  );
};

export default ImageList;
