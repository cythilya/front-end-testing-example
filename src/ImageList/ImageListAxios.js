import { useEffect, useState } from 'react';
import axios from 'axios';
import ImageItem from './ImageItem';

const ImageList = () => {
  const [images, setImages] = useState([]);

  useEffect(() => {
    async function fetchData() {
      try {
        const response = await axios.get('https://dummyjson.com/products');
        setImages(response.data.products);
      } catch (error) {
        throw error;
      }
    }
    fetchData();
  }, []);

  const renderImages = () => {
    return images.map(({ id, title, images }) => {
      return <ImageItem key={id} title={title} image={images[0]} />;
    });
  };

  return (
    <>
      {images.length !== 0 ? (
        renderImages()
      ) : (
        <div data-test-id="no-data-prompt">No data to display.</div>
      )}
    </>
  );
};

export default ImageList;
