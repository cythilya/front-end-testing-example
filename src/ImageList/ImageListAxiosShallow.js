import { useState } from 'react';
import ImageItem from './ImageItem';

const ImageList = (data) => {
  const [images, setImages] = useState(data.products);

  const renderImages = () => {
    return images.map(({ id, title, images }) => (
      <ImageItem key={id} title={title} image={images[0]} />
    ));
  };

  const renderNoDataPrompt = () => {
    return <div data-test-id="no-data-prompt">No data to display.</div>;
  };

  return <>{!!images.length ? renderImages() : renderNoDataPrompt()}</>;
};

export default ImageList;
