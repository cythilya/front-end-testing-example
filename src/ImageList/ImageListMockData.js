import { useEffect, useState } from 'react';
import axios from 'axios';
import ImageItem from './ImageItem';

const ImageList = ({ data }) => {
  const renderImages = () => {
    return data.products.map(({ id, title, images }) => (
      <ImageItem key={id} title={title} image={images[0]} />
    ));
  };

  const renderNoDataPrompt = () => {
    return <div data-test-id="no-data-prompt">No data to display.</div>;
  };

  return <>{!!data.products.length ? renderImages() : renderNoDataPrompt()}</>;
};

export default ImageList;
