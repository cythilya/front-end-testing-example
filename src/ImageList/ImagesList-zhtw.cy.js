import ImageItem from './ImageItem';

describe('ImageList', () => {
  describe('ImageList', () => {
    it('should render no data prompt when no images', () => {
      // 攔截 API https://dummyjson.com/products 的回應，使用 fixtures 資料夾中的 `empty.json`
      // 渲染 `<ImageList />` 元件
      // 斷言沒有資料時會出現 "no-data-prompt" 提示
    });

    it('should render correct image items when have data', () => {
      // 攔截 API https://dummyjson.com/products 的回應，使用 fixtures 資料夾中的 `empty.json`
      // 渲染 `<ImageList />` 元件
      // 斷言 image items length is 2
    });
  });
});
