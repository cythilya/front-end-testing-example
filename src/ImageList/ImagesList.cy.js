import ImageList from './ImageList';

describe('ImageList', () => {
  it('should render no data prompt when no images', () => {
    cy.intercept('GET', 'https://dummyjson.com/products', {
      statusCode: 200,
      body: { products: [] },
    }).as('apiRequest');

    cy.mount(<ImageList />);

    cy.get('[data-test-id="no-data-prompt"]').should('exist');
  });

  it('should render correct image items when have data', () => {
    cy.intercept('GET', 'https://dummyjson.com/products', {
      statusCode: 200,
      body: {
        products: [
          {
            id: '1',
            title: '打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵',
            images: ['https://bit.ly/4228IT0'],
          },
          {
            id: '2',
            title: '島波海道。單車跳島之旅',
            images: ['https://bit.ly/4b5nLPO'],
          },
        ],
      },
    }).as('apiRequest');

    cy.mount(<ImageList />);

    cy.get('[data-test-id="image-item-title"]').should('have.length', 2);
    cy.contains('打造高速網站');
    cy.contains('島波海道');
  });
});
