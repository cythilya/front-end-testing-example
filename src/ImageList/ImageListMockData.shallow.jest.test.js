import React from 'react';
import { act } from 'react-test-renderer';
import { createRenderer } from 'react-test-renderer/shallow';
import ImageList from './ImageListMockData';

describe('ImageList', () => {
  it('should render no data prompt when no images', async () => {
    const renderer = createRenderer();
    const mockedData = { products: [] };

    const wrapper = await act(async () => renderer.render(<ImageList data={mockedData}/>));

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correct image items when have data', async () => {
    const renderer = createRenderer();
    const mockedData = {
      products: [
        {
          id: '1',
          title:
            '打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵',
          images: ['https://bit.ly/4228IT0'],
        },
        {
          id: '2',
          title: '島波海道。單車跳島之旅',
          images: ['https://bit.ly/4b5nLPO'],
        },
      ],
    };

    const wrapper = await act(async () => renderer.render(<ImageList data={mockedData} />));

    expect(wrapper).toMatchSnapshot();
  });
});
