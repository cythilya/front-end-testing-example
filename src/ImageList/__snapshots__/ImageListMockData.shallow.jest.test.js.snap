// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageList should render correct image items when have data 1`] = `
<React.Fragment>
  <ImageItem
    image="https://bit.ly/4228IT0"
    title="打造高速網站，從網站指標開始！全方位提升使用者體驗與流量的關鍵"
  />
  <ImageItem
    image="https://bit.ly/4b5nLPO"
    title="島波海道。單車跳島之旅"
  />
</React.Fragment>
`;

exports[`ImageList should render no data prompt when no images 1`] = `
<React.Fragment>
  <div
    data-test-id="no-data-prompt"
  >
    No data to display.
  </div>
</React.Fragment>
`;
